src/dune_rules/mdx.ml

```diff
@@ -182,36 +199,123 @@ let gen_rules_for_single_file stanza ~sctx ~dir ~expander ~mdx_prog src =
+
+    Build.with_no_targets
+      (Dep_conf_eval.unnamed ~expander (pkg_deps @ other_deps))
     >>> Build.with_no_targets (Build.dyn_deps dyn_deps)
-    >>> Command.run ~dir:(Path.build dir) mdx_prog
-          ( [ Command.Args.A "test" ] @ prelude_args
-          @ [ A "-o"; Target files.corrected; Dep (Path.build files.src) ] )
+    >>> Command.run ~dir:(Path.build dir) ~stdout_to:files.corrected executable
+          command_line
   in
+
   Super_context.add_rule sctx ~loc ~dir mdx_action;
   (* Attach the diff action to the @runtest for the src and corrected files *)
   let diff_action = Files.diff_action files in
   Super_context.add_alias_action sctx (Alias.runtest ~dir) ~loc:(Some loc) ~dir
     ~stamp:("mdx", files.src)
     (Build.with_no_targets diff_action)
 
+let name = "mdx_gen"
+
+let mdx_prog_gen t ~sctx ~dir ~scope ~expander ~mdx_prog =
+  let loc = t.loc in
+  let dune_version = Scope.project scope |> Dune_project.dune_version in
+  let file = Path.Build.relative dir "mdx_gen.ml-gen" in
+
+  (* Libs from the libraries field should have their include directories sent to
+     mdx *)
+  let libs_to_include =
+    List.filter_map t.libraries ~f:(function
+      | Direct lib
+      | Re_export lib ->
+        Result.to_option (Lib.DB.resolve (Scope.libs scope) lib)
+      | _ -> None)
+  in
+
+  let libs_include_paths = Lib.L.include_paths libs_to_include in
+
+  let directory_args =
+    Path.Set.to_list libs_include_paths
+    |> List.map ~f:(fun p ->
+           Command.Args.[ A "--directory"; A (Path.to_absolute_filename p) ])
+    |> List.flatten
+  in
+
+  let prelude_args = List.concat_map t.preludes ~f:(Prelude.to_args ~dir) in
+
+  (* We call mdx to generate the testing executable source *)
+  let action =
+    Command.run ~dir:(Path.build dir) mdx_prog ~stdout_to:file
+      ((Command.Args.A "dune-gen" :: prelude_args) @ directory_args)
+  in
+  Super_context.add_rule sctx ~loc ~dir action;
+
+  (* We build the generated executable linking in the libs from the libraries
+     field *)
+  let obj_dir = Obj_dir.make_exe ~dir ~name in
+  let main_module_name = Module_name.of_string name in
+  let module_ = Module.generated ~src_dir:(Path.build dir) main_module_name in
+  let modules = Modules.singleton_exe module_ in
+  let flags = Ocaml_flags.default ~dune_version ~profile:Release in
+  let compile_info =
+    Lib.DB.resolve_user_written_deps_for_exes (Scope.libs scope)
+      [ (t.loc, name) ]
+      (Lib_dep.Direct (loc, Lib_name.of_string "mdx.test") :: t.libraries)
+      ~pps:[ (*Preprocess.Per_module.pps t.preprocess*) ] ~dune_version
+      ~optional:false
+  in
+  let cctx =
+    Compilation_context.create ~super_context:sctx ~scope ~expander ~obj_dir
+      ~modules ~flags
+      ~requires_compile:(Lib.Compile.direct_requires compile_info)
+      ~requires_link:(Lib.Compile.requires_link compile_info)
+      ~opaque:(Explicit false) ~js_of_ocaml:None ~dynlink:false ~package:None ()
```

Comment:
```suggestion
      ~opaque:(Explicit false) ~js_of_ocaml:None ~dynlink:true ~package:None ()
```

See #4081 - this might not work on arm32 otherwise.